cri/v1alpha2/cri.go

```diff
@@ -246,12 +246,29 @@ func (c *CriManager) RunPodSandbox(ctx context.Context, r *runtime.RunPodSandbox
 		return nil, fmt.Errorf("failed to create a sandbox for pod %q: %v", config.Metadata.Name, err)
 	}
 	id := createResp.ID
+
+	// once sandbox container created, we are obligated to store it.
+	sandboxMeta := &SandboxMeta{
+		ID:      id,
+		Config:  config,
+		Runtime: config.Annotations[anno.KubernetesRuntime],
+	}
+	if err := c.SandboxStore.Put(sandboxMeta); err != nil {
+		return nil, err
+	}
+
+	// If running sandbox failed, clean up the container.
 	defer func() {
-		// If running sandbox failed, clean up the container.
 		if retErr != nil {
 			err := c.ContainerMgr.Remove(ctx, id, &apitypes.ContainerRemoveOptions{Volumes: true, Force: true})
 			if err != nil {
-				logrus.Errorf("failed to remove the container when running sandbox failed: %v", err)
+				logrus.Errorf("failed to remove container when running sandbox failed %q: %v", id, err)
+			}
+			// should not remove the sandbox container metadata from sandboxStore
+			// until it was removed by pouchd.
+			removeErr := c.SandboxStore.Remove(id)
```

Comment:
use `if err := xx; err != nil pattern` here?

Reply:
Yes, it is better to use `if err := xx; err != nil`, and the err would only take effect in the code block. After that, there is no need to add a new naming `removeErr`, `err` is enough.